sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3071,15 +3071,31 @@ class Analyzer(
       case p => p transformExpressions {
         case u @ UpCast(child, _, _) if !child.resolved => u
 
-        case UpCast(child, dt: AtomicType, _)
+        case UpCast(_, target, _) if target != DecimalType && !target.isInstanceOf[DataType] =>
+          throw new AnalysisException(
+            s"UpCast only support DecimalType as AbstractDataType yet, but got: $target")
+
+        case UpCast(child, target, walkedTypePath) if target == DecimalType
+          && child.dataType.isInstanceOf[DecimalType] =>
+          assert(walkedTypePath.nonEmpty,
+            "object DecimalType should only be used inside ExpressionEncoder")
+          // SPARK-31750: for the case where data type is explicitly known, e.g, spark.read
```

Comment:
nit:
```
SPARK-31750: if we want to upcast to the general decimal type, and the `child` is already
decimal type, we can remove the `Upcast` and accept any precision/scale.
This can happen for cases like `spark.read.parquet("/tmp/file").as[BigDecimal]`.
```